### PR TITLE
adding foreach support for ForwardRanges [FEEDBACK WANTED]

### PR DESCRIPTION
**Pull request by UplinkCoder**
another syntatic-sugar feature done.
shuffling a lot of code around ... this sure looks ugly...

### CODE REVIEW

**Comment by deadalnix** [comment]
Explainaining what you did here would certainly be valuable for review.

**Comment by UplinkCoder** [comment]
Well it's rather easy.
IF a struct or a class defines the members "front","empty" and "popFront"
`handleRange()` generates a ForLoop like this 
```
for(auto index = typeof(range,front()).init; !range.empty();range.popFront()) {
    index = range.front;
    // foreach body
}
```

File: src/d/semantic/statement.d [hunk 1]
```diff
@@ -272,82 +272,131 @@ public:
...
+			
+			flattenedStmts ~= new ForStatement(loc, initialize, condition, increment, stmt);
+		} else if (iType.kind == TypeKind.Class) {
+			if (handleRange(iType.dclass)) {return;}
+			assert(0, "foreach for classes does not work yet");
+		} else if (iType.kind == TypeKind.Struct) {
+			if(handleRange(iType.dstruct)) {return;} 
+			else assert(0, "foreach for structs does not work yet");
+		} else {
+			assert(0, "you can't do foreach on this at all");
```
Comment (schancel) [inline src/d/semantic/statement.d]: Please make these proper compiler errors so they give good output, and fix the wording.  Don't use "you," or contractions in the error.   Also, we should consider instrumenting an i18n library so we can localize SDC eventually.

**Comment by schancel** [comment]
This is a good thing to get working.  Can you please finish this up @UplinkCoder ?

**Pull request closed without merging**